### PR TITLE
Added support for running T2 applicable tests using vsonic(KVM) for n…

### PR DESCRIPTION
**Pull request by mannytaheri**
### Description of PR
Added support for running T2 applicable tests using vsonic(KVM) for neighbors

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205

#### What is the motivation for this PR?
To add support for vsonic/KVM commands on neighbors, as the commands are different from Eo

#### How did you do it?
1. Fix for test_lldp testcase:
    KVM neighbors do not advertise port ifname. They instead advertise 'descr'.  We will use ifname for EosHost and descr for KVM

2. Fix for test_snmp_loopback testcase:
    snmpget command requires running in snmp docker

    Eos Example:
          ARISTA01T3#bash snmpget -v2c -c public 10.1.0.1 1.3.6.1.2.1.1.1.0
         SNMPv2-MIB::sysDescr.0 = STRING: SONiC Software Version: SONiC.HEAD.291231-msft-2205-c679f2d - HwSku: Nokia- 
         IXR7250E-36x100G - Distribution: Debian 11.4 - Kernel: 5.10.0-12-2-amd64

    vSonic Exmple:
        admin@ARISTA01T1:~$ docker exec -it snmp snmpget -v2c -c public 10.1.0.1 1.3.6.1.2.1.1.1.0
        iso.3.6.1.2.1.1.1.0 = STRING: "SONiC Software Version: SONiC.HEAD.291231-msft-2205-c679f2d - HwSku: Nokia-IXR7250E- 
       6x400G - Distribution: Debian 11.4 - Kernel: 5.10.0-12-2-amd64"

3. Fix for  voq tests
    BGP:
        - Use vtysh for BGP configuration on vsonic 
    Interface up/down:
        - Use shut/no shut commands
    
4.  voq_helper
    Interface name and numbering:
       - Neighbors interface name is difference on vsonic vs Eos. 
           - On Eos, inerfaces are named 'eth' (Ex. eth1) and portchannels are named po (Ex. po1).
           - On vsonic, interfaces are named Ethernet (Ex. Ethernet1) and portchannels are named PortChannel (PortChannel1).

5. Fix for test_lag interface
    There is no support for setting lacp rate in vsonic. So, skipping these tests.

#### How did you verify/test it?
Tested the code on a dut that uses Eos VM as it's neighbors in one setup and against vsonic neighbors in another setup.

### CODE REVIEW

File: tests/voq/test_voq_nbr.py [hunk 1]
```diff
@@ -128,16 +129,19 @@ def restore_bgp(duthosts, nbrhosts, all_cfg_facts):
...
+                                if ":" in address:
+                                    nbr['host'].eos_config(
+                                        lines=["no ipv6 route ::/0 %s " % neighbor])
+                                else:
+                                    nbr['host'].eos_config(
+                                        lines=["no ip route 0.0.0.0/0 %s " % neighbor])
                             else:
-                                nbr['host'].eos_config(
-                                    lines=["no ip route 0.0.0.0/0 %s " % neighbor])
+                                nbr['host'].shell("sudo vtysh -c 'configure terminal' -c 'router bgp " + str(nbr['conf']['bgp']['asn']) + "' -c 'no neighbor {} shutdown'".format(neighbor))
```
Comment (arlakshm) [inline tests/voq/test_voq_nbr.py:144]: we don't need to add default route if its sonic devices?

File: tests/voq/test_voq_nbr.py [hunk 2]
```diff
@@ -220,21 +224,24 @@ def disable_nbr_bgp_neighs(node=None, results=None):
...
+                        node_results.append(node['host'].eos_config(
+                            lines=["ip route 0.0.0.0/0 %s " % neighbor],
+                            module_ignore_errors=True)
+                        )
                 else:
-                    node_results.append(node['host'].eos_config(
-                        lines=["ip route 0.0.0.0/0 %s " % neighbor],
-                        module_ignore_errors=True)
-                    )
+                    node_results.append(node['host'].shell("sudo vtysh -c 'configure terminal' -c 'router bgp " + str(node['conf']['bgp']['asn'])+ "' -c 'neighbor {} shutdown'".format(neighbor)))
```
Comment (arlakshm) [inline tests/voq/test_voq_nbr.py:244]: we don't need to add default route if its sonic device?

File: tests/voq/test_voq_nbr.py [hunk 3]
```diff
@@ -436,12 +450,25 @@ def _change_vm_interface_on_vm(nbrhosts, state="up", node=None, results=None):
         for eos_intf in nbr['conf']['interfaces'].keys():
             if "Loopback" in eos_intf:
                 continue
+
             if state == "up":
-                logger.info("Startup EOS %s interface %s", node, eos_intf)
-                node_results.append(nbr['host'].eos_config(lines=["no shutdown"], parents=["interface %s" % eos_intf]))
+                logger.info("Startup Nbr %s interface %s", node, eos_intf)
```
Comment (arlakshm) [inline tests/voq/test_voq_nbr.py:455]: can we use the Fanout class functions to shutdown and startup interface iso doing if/else here
https://github.com/sonic-net/sonic-mgmt/blob/master/tests/common/devices/fanout.py has these function defined for different type of fanout devices

**Review by arlakshm — COMMENTED** [review]
as comments

File: tests/voq/test_voq_nbr.py (hunk 3 above)
Comment (mannytaheri) [reply tests/voq/test_voq_nbr.py:455]: The function 'change_vm_interfaces' is used in tests test_neighbor_clear_one and test_neighbor_clear_all. The intent of these test is to clear the neighbor without bringing the actual port on the linecard down. So, bringing the fanout port down, changes the intent of the test entirely.

File: tests/voq/test_voq_nbr.py (hunk 1 above)
Comment (mannytaheri) [reply tests/voq/test_voq_nbr.py:137]: Removed it from eos.

File: tests/voq/test_voq_nbr.py (hunk 2 above)
Comment (mannytaheri) [reply tests/voq/test_voq_nbr.py:227]: Removed it from eos.

**Comment by abdosi** [comment]
cc @judyjoseph

**Comment by abdosi** [comment]
@sanmalho-git Can we close this soon.

**Comment by wangxin** [comment]
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The detected issues may be old or new. For new issues, please try to fix them.

For old issues, it is not mandatory to fix them because they were not caused by this change. It is unfair to blame
author of this pull request. But if you can take extra effort to fix the old issues as well, that would be great!

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Failed<br/>- hook id: trailing-whitespace<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/pc/test_lag_2.py<br/>Fixing tests/lldp/test_lldp.py<br/><br/>fix end of files.........................................................Failed<br/>- hook id: end-of-file-fixer<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/snmp/test_snmp_loopback.py<br/>Fixing tests/voq/test_voq_nbr.py<br/><br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>...<br/>[truncated extra lines, please run pre-commit locally to view full check results]<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

**Comment by mrkcmo** [comment]
Which topology was this ran in for this setup? Was it the t2-topo?

**Comment by mannytaheri** [comment]
> Which topology was this ran in for this setup? Was it the t2-topo?

T2 Toplogy

**Comment by mrkcmo** [comment]
> > Which topology was this ran in for this setup? Was it the t2-topo?
> 
> T2 Toplogy

Thanks. Was this a physical T2 setup with vSONIC KVM neighbors?

**Comment by mannytaheri** [comment]
> > > Which topology was this ran in for this setup? Was it the t2-topo?
> > 
> > 
> > T2 Toplogy
> 
> Thanks. Was this a physical T2 setup with vSONIC KVM neighbors?

The topology that we are using is [T2_min](https://github.com/sonic-net/sonic-mgmt/blob/master/ansible/vars/topo_t2_2lc_min_ports-masic.yml) with a physical chassis with 2 Linecards and 1 supervisor card and all eBGP peers being simulated using vsonic

File: tests/snmp/test_snmp_loopback.py [hunk 4]
```diff
@@ -66,8 +70,11 @@ def test_snmp_loopback(duthosts, enum_rand_one_per_hwsku_frontend_hostname,
...
-            "Sysdescr not found in SNMP result from IP {}".format(ip)
-        assert snmp_facts['ansible_sysdescr'] in result['stdout_lines'][0][
-            0], "Sysdescr from IP{} not matching with result from Mgmt IPv4.".format(ip)
+        assert len(result[u'stdout_lines']) > 0, 'No result from snmpget'
+
+        if isinstance(nbr['host'], EosHost):
+            assert "SONiC Software Version" in result[u'stdout_lines'][0][0], "Sysdescr not found in SNMP result from IP {}".format(ip)
+            assert snmp_facts['ansible_sysdescr'] in result[u'stdout_lines'][0][0], "Sysdescr from IP{} not matching with result from Mgmt IPv4.".format(ip)
+        else:
+            assert "SONiC Software Version" in result[u'stdout_lines'][0], "Sysdescr not found in SNMP result from IP {}".format(ip)
```
Comment (SuvarnaMeenakshi) [inline tests/snmp/test_snmp_loopback.py:79]: the assert is checking for the result.
Could modify get_snmp_output() to get the result in a uniform format so that the test function need not check for the type of nbr host.

File: tests/lldp/test_lldp.py [hunk 5]
```diff
@@ -33,11 +33,15 @@ def test_lldp(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost,
         # Compare the LLDP neighbor name with minigraph neigbhor name (exclude the management port)
         assert v['chassis']['name'] == config_facts['DEVICE_NEIGHBOR'][k]['name']
         # Compare the LLDP neighbor interface with minigraph neigbhor interface (exclude the management port)
-        assert v['port']['ifname'] == config_facts['DEVICE_NEIGHBOR'][k]['port']
+        if request.config.getoption("--neighbor_type") == 'eos':
```
Comment (SuvarnaMeenakshi) [inline tests/lldp/test_lldp.py:36]: Will the default value for --neighbor_type be 'eos' if we do not pass this option in pytest?

File: tests/lldp/test_lldp.py [hunk 6]
```diff
@@ -33,11 +33,15 @@ def test_lldp(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost,
...
+            assert v['port']['ifname'] == config_facts['DEVICE_NEIGHBOR'][k]['port']
+        else:
+            # Dealing with KVM that advertises port description
+            assert v['port']['descr'] == config_facts['DEVICE_NEIGHBOR'][k]['port']
 
 
-def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost, eos,
-                       collect_techsupport_all_duts, loganalyzer, enum_frontend_asic_index, tbinfo):
+def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost, eos, sonic,
```
Comment (SuvarnaMeenakshi) [inline tests/lldp/test_lldp.py:43]: Why do we need the additional parameter sonic?

File: tests/snmp/test_snmp_loopback.py [hunk 7]
```diff
@@ -25,9 +27,15 @@ def get_snmp_output(ip, duthost, nbr, creds_all_duts):
         iptables_cmd, ip)
     duthost.shell(ip_tbl_rule_add)
 
-    eos_snmpget = "bash snmpget -v2c -c {} {} 1.3.6.1.2.1.1.1.0".format(
-        quote(creds_all_duts[duthost.hostname]['snmp_rocommunity']), ip)
-    out = nbr['host'].eos_command(commands=[eos_snmpget])
+    if isinstance(nbr['host'], EosHost):
```
Comment (SuvarnaMeenakshi) [inline tests/snmp/test_snmp_loopback.py:30]: I have another sonic-mgmt PR which will handle this change https://github.com/sonic-net/sonic-mgmt/pull/8802 
You could remove these changes from test_snmp_loopback.py

File: tests/voq/test_voq_nbr.py [hunk 8]
```diff
@@ -219,21 +217,25 @@ def disable_nbr_bgp_neighs(node=None, results=None):
...
                         module_ignore_errors=True)
                     )
+                    if ":" in neighbor:
+                        node_results.append(node['host'].eos_config(
+                            lines=["ipv6 route ::/0 %s " % neighbor],
+                            module_ignore_errors=True)
+                        )
+                    else:
+                        node_results.append(node['host'].eos_config(
+                            lines=["ip route 0.0.0.0/0 %s " % neighbor],
```
Comment (SuvarnaMeenakshi) [inline tests/voq/test_voq_nbr.py:233]: Why the change from ::/0 to IPv4?

File: tests/voq/test_voq_nbr.py [hunk 9]
```diff
@@ -219,21 +217,25 @@ def disable_nbr_bgp_neighs(node=None, results=None):
...
+                            module_ignore_errors=True)
+                        )
                 else:
-                    node_results.append(node['host'].eos_config(
-                        lines=["ip route 0.0.0.0/0 %s " % neighbor],
-                        module_ignore_errors=True)
-                    )
+                    node_results.append(node['host'].shell("sudo vtysh -c 'configure terminal' -c 'router bgp " + str(
```
Comment (SuvarnaMeenakshi) [inline tests/voq/test_voq_nbr.py:237]: for eos neighbor, there is a check added IPv6 neighbor and is handled in a different way, do we need that for vsonic as well?

File: tests/voq/voq_helpers.py [hunk 10]
```diff
@@ -410,12 +411,20 @@ def get_eos_mac(nbr, nbr_intf):
...
+            # convert Port-Channel1 to po1
+            shell_intf = "po" + nbr_intf[-1]
+        else:
+            # convert Ethernet1 to eth1
+            shell_intf = "eth" + nbr_intf[-1]
     else:
-        # convert Ethernet1 to eth1
-        shell_intf = "eth" + nbr_intf[-1]
+        if "port-channel" in nbr_intf.lower():
+            # convert Port-Channel1 to po1
```
Comment (SuvarnaMeenakshi) [inline tests/voq/voq_helpers.py:424]: In the comment, should change po1 to PortChannel1

File: tests/voq/test_voq_nbr.py (hunk 3 above)
Comment (SuvarnaMeenakshi) [reply tests/voq/test_voq_nbr.py:453]: @arlakshm can you help review test_voq_nbr changes

**Comment by rlhui** [comment]
Testing of this is pending https://github.com/sonic-net/sonic-mgmt/pull/8418

**Comment by SuvarnaMeenakshi** [comment]
@mannytaheri  https://github.com/sonic-net/sonic-mgmt/pull/8418 is merged in 202205, will you be able to test and update this PR as per comments and resolve merge conflicts

File: ansible/roles/test/files/tools/loganalyzer/loganalyzer.py [hunk 11]
```diff
@@ -767,10 +767,8 @@ def main(argv):
     verbose = False
 
     try:
-        opts, args = getopt.getopt(argv, "a:r:s:l:o:m:i:e:vh",
-                                   ["action=", "run_id=", "start_marker=", "logs=",
-                                    "out_dir=", "match_files_in=", "ignore_files_in=",
-                                    "expect_files_in=", "verbose", "help"])
+        opts, args = getopt.getopt(argv, "a:r:s:l:o:m:i:e:vh", [
+                                   "action=", "run_id=", "start_marker=", "logs=", "out_dir=", "match_files_in=", "ignore_files_in=", "expect_files_in=", "verbose", "help"])
```
Comment (judyjoseph) [inline ansible/roles/test/files/tools/loganalyzer/loganalyzer.py:771]: @mannytaheri Can you update this here -- the line is too long and build fails with the following error

ansible/roles/test/files/tools/loganalyzer/loganalyzer.py:771:121: E501 line too long (173 > 120 characters)

**Review by SuvarnaMeenakshi — APPROVED** [review]
lgtm, Can you merge with latest with conflict resolved.
I also have a minor comment open.

File: tests/voq/voq_helpers.py (hunk 10 above)
Comment (mannytaheri) [reply tests/voq/voq_helpers.py]: Fixed

File: tests/voq/test_voq_nbr.py (hunk 9 above)
Comment (mannytaheri) [reply tests/voq/test_voq_nbr.py:237]: vsonic is working fine without the change.

File: tests/snmp/test_snmp_loopback.py (hunk 7 above)
Comment (mannytaheri) [reply tests/snmp/test_snmp_loopback.py]: Fixed

File: tests/lldp/test_lldp.py (hunk 6 above)
Comment (mannytaheri) [reply tests/lldp/test_lldp.py:43]: Both eos and sonic are needed to read and yield eos/sonic configuration.

File: tests/lldp/test_lldp.py (hunk 5 above)
Comment (mannytaheri) [reply tests/lldp/test_lldp.py:36]: Yes, the default is eos

**Comment by SuvarnaMeenakshi** [comment]
@mannytaheri  failure seen at sonic-mgmt PR checker , fails at pytest collection with below error:
____________________ ERROR collecting voq/test_voq_intfs.py ____________________
ImportError while importing test module '/var/src/sonic-mgmt/tests/voq/test_voq_intfs.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
voq/test_voq_intfs.py:13: in <module>
    from .test_voq_disrupts import check_bgp_neighbors
voq/test_voq_disrupts.py:8: in <module>
    from .test_voq_ipfwd import pick_ports
voq/test_voq_ipfwd.py:20: in <module>
    from .test_voq_nbr import LinkFlap
voq/test_voq_nbr.py:18: in <module>
    from voq_helpers import get_neighbor_info
E   ModuleNotFoundError: No module named 'voq_helpers'

**Comment by mannytaheri** [comment]
> @mannytaheri failure seen at sonic-mgmt PR checker , fails at pytest collection with below error: ____________________ ERROR collecting voq/test_voq_intfs.py ____________________ ImportError while importing test module '/var/src/sonic-mgmt/tests/voq/test_voq_intfs.py'. Hint: make sure your test modules/packages have valid Python names. Traceback: /usr/lib/python3.8/importlib/**init**.py:127: in import_module return _bootstrap._gcd_import(name[level:], package, level) voq/test_voq_intfs.py:13: in from .test_voq_disrupts import check_bgp_neighbors voq/test_voq_disrupts.py:8: in from .test_voq_ipfwd import pick_ports voq/test_voq_ipfwd.py:20: in from .test_voq_nbr import LinkFlap voq/test_voq_nbr.py:18: in from voq_helpers import get_neighbor_info E ModuleNotFoundError: No module named 'voq_helpers'

@SuvarnaMeenakshi  @judyjoseph  I have fixed the PR.  All testcases passed.  Please help to merge it if it looks ok.

**Comment by mannytaheri** [comment]
@SuvarnaMeenakshi  @judyjoseph  I have fixed the PR. All testcases passed. Please help to merge it if it looks ok.

File: ansible/roles/test/files/tools/loganalyzer/loganalyzer.py [hunk 12]
```diff
@@ -219,28 +220,30 @@ def wait_for_marker(self, marker, timeout=120, polling_interval=10):
...
+            with open(syslog_file, 'r') as fp:
+                dt = os.path.getctime(syslog_file)
+                if last_dt != dt:
+                    try:
+                        with open(prev_syslog_file, 'r') as pfp:
+                            pfp.seek(last_check_pos)
+                            for log in fp:
+                                if marker in log:
+                                    return True
+                    except OSError:
```
Comment (judyjoseph) [inline ansible/roles/test/files/tools/loganalyzer/loganalyzer.py:234]: @mannytaheri can you share why these changes are needed here in loganalyzer ? This is common code which gets used for non vsonic tests also.

**Comment by mannytaheri** [comment]
@judyjoseph This PR is 18 months old.  The loganalyzer changes were suggested by Sandeep and the changes have been working for both vsonic and non vsonic on our chassis.  If you think we don't need them I can remove them and test with both vsonic and non vsonic?

**Review by SuvarnaMeenakshi — APPROVED** [review]
lgtm

**Review by judyjoseph — APPROVED** [review]
